documentation/TagAnalysis.md

```diff
@@ -0,0 +1,187 @@
+# Tag Analysis
+
+## Overview
+
+Tag analysis is a compile-time static analysis that keeps track of information flows in programs. Information is said
+to flow from object X to object Y, whenever information stored in X is transferred to object Y. Tag analysis attaches
+different kinds of tags to objects in a program to keep track of dataflows for these tags. For example, taint analysis
+is a special form of tag analysis that reasons about sensitive information, such as private user data in social network
+service, or private keys in crypto code. Using tag analysis, a static analyzer can verify that tainted objects (i.e.,
+objects attached with special taint tags) will never flow to specific program locations that may leak the information,
+e.g., a method writing the tainted objects to a public channel. Because the analysis is carried out at compile time,
+the analyzer can verify non-trivial information-flow properties without any runtime overheads.
+
+The tag analysis provided by MIRAI is context-, field-, flow-, and path-sensitive. Context-sensitivity and field-
+sensitivity come from MIRAI's summary-based inter-procedural analysis and precise abstract heap model, respectively.
+The tag analysis is implemented as a dataflow analysis with a tag domain that over-approximates present and absent tags
+on program objects simultaneously. When the dataflow analysis is not precise enough to answer queries, MIRAI generates
+a propositional formula that encodes both the query and the path condition that guards the query, and uses an
+off-the-shelf constraint solver such as [Z3](https://github.com/Z3Prover/z3) to answer the query.
+
+Tag analysis is in general an undecidable problem, so it might not obtain sufficiently precise information to verify
+information-flow properties. The tag analysis provided by MIRAI is developed to consist with MIRAI's design, which
+avoids false negatives (not flagging real errors), and reduces the number of false positives (flagging false errors).
```

Comment:
Since this point seems particularly important, I would love to read a short explanation of the explicit link between this, positive/negative taint analysis, and timeouts.
I think you already have the main elements for that (i.e. examples where you check presence/absence of a tag). What's missing, I think, is one sentence here on how timeouts might occur when the tracking blows up, and once sentence per example below laying out the "in situ" consequence: "here, if the data flow was *really* complex, MIRAI would be unable to check for the presence of the tag and would report an error"